dtm0/dtx.h

```diff
@@ -0,0 +1,156 @@
+	 * This DTX was created at the beginning of the time, and it was
+	 * propagated all over the codebase. Since it is really hard to
+	 * remove it, we instead venerate it by providing the very first
+	 * place in this new dtm0 structure. It helps to have a simple
+	 * typecast (without branded object) and no extra allocations for
+	 * this ancient but not really useful at this moment structure.
+	 */
+	struct m0_dtx           dd_ancient_dtx;
+	/* see m0_dtm0_dtx_state */
+	struct m0_sm            dd_sm;
+	struct m0_dtm0_tx_desc  dd_txd;
+	struct m0_dtm0_service *dd_dtms;
+	uint32_t                dd_nr_executed;
+	struct m0_sm_ast        dd_exec_all_ast;
+
+	/* XXX: The implementation is very simple and it relies on the idea
+	 * of fully replicated requests. Therefore, all FOP submitted
+	 * as a part of this DTX should to be identical (at least
+	 * from the REDO mechanism standpoint). So that we do not use
+	 * any lists here. But we will have to some time later.
+	 */
+	const struct m0_fop    *dd_fop;
+};
+
+/** An AST object for persistent message processing. */
+struct m0_dtm0_pmsg_ast {
+	/** AST where the update is applied */
+	struct m0_sm_ast       p_ast;
+	/** A FOP that contains a partial update to be applied. */
+	struct m0_fop         *p_fop;
+	/** A dtx that has to be update */
+	struct m0_dtm0_dtx    *p_dtx;
+};
+
+M0_INTERNAL void m0_dtm0_dtx_domain_init(void);
+M0_INTERNAL void m0_dtm0_dtx_domain_fini(void);
+
+/* The API below extends the existing m0_dtx API. Since it operates
+ * on m0_dtx structure in the "DTM0"-way, the naming convention here is
+ * a bit different (m0_dtx + dtm0 => m0_dtx0).
+ */
+
+M0_INTERNAL struct m0_dtx* m0_dtx0_alloc(struct m0_dtm0_service *svc,
+					 struct m0_sm_group     *grp);
+
+/** Assigns a TID to the transaction. */
+M0_INTERNAL int m0_dtx0_prepare(struct m0_dtx *dtx);
+
+M0_INTERNAL int m0_dtx0_open(struct m0_dtx  *dtx, uint32_t nr);
+
+/** Fills in the FID of the given participant.
+ * @param pa_idx  A position in the participants list.
+ * @param pa_sfid The FID of a service that participates in the dtx.
+ */
+M0_INTERNAL int m0_dtx0_fid_assign(struct m0_dtx       *dtx,
+				   uint32_t             pa_idx,
+				   const struct m0_fid *pa_sfid);
+
+/** Fills in the FOP associated with the given participant. */
+M0_INTERNAL void m0_dtx0_fop_assign(struct m0_dtx       *dtx,
+				    uint32_t             pa_idx,
+				    const struct m0_fop *pa_fop);
+M0_INTERNAL int m0_dtx0_close(struct m0_dtx *dtx);
+
+M0_INTERNAL void m0_dtx0_executed(struct m0_dtx *dtx, uint32_t pa_idx);
+
+/** Marks a transaction as "no longer in-use".
+ * The user does not have exclusive ownership on a dtx after it has
+ * been closed (i.e., added to the log). This function ends this
+ * sharing, so that the log has exclusive ownership of the record.
+ * The user should not hold any pointers to this DTX after this call.
+ */
+M0_INTERNAL void m0_dtx0_done(struct m0_dtx *dtx);
```

Comment:
Please clarify the calling order of "m0_dtx0_close" and "m0_dtx0_done".
Which one should we call first?

Reply:
m0_dtx0_close shall be called when the user of DTM0 finished populating the information about the participants of the corresponding dtx. Once a dtx is closed, all the participants in the list will move to INPROGRESS state.
m0_dtx0_done shall be called when the user of DTM0 is ready to release the resources associated with the dtx. Note: dtx0_done() may not destroy the dtx (hence, it is not "fini" but "done"), however the user must invalidate any references to this dtx.

Added these comments to the header.